packages/nextjs/src/config/wrappers/withSentryAPI.ts

```diff
@@ -0,0 +1,45 @@
+// We import these types from `withSentry` rather than directly from `next` because our version can work simultaneously
+// with multiple versions of next. See note in `withSentry` for more.
+import type { NextApiHandler, WrappedNextApiHandler } from '../../utils/withSentry';
+import { withSentry } from '../../utils/withSentry';
+
+/**
+ * Wrap the given API route handler for tracing and error capturing. Thin wrapper around `withSentry`, which only
+ * applies it if it hasn't already been applied.
+ *
+ * @param maybeWrappedHandler The handler exported from the user's API page route file, which may or may not already be
+ * wrapped with `withSentry`
+ * @param parameterizedRoute The page's route, passed in via the proxy loader
+ * @returns
+ */
+export function withSentryAPI(
+  maybeWrappedHandler: NextApiHandler | WrappedNextApiHandler,
+  parameterizedRoute: string,
+): WrappedNextApiHandler {
+  // We want the innards of `withSentry` to have access to the parameterized route, so it can be used when we start the
+  // request transaction. If we were always the ones calling `withSentry` (the way we're always the ones to call
+  // `withSentryServerSideProps`, for example), then we could just pass it in as a second parameter and know it would
+  // always be there. But in the case where users have already manually wrapped their API route handlers with
+  // `withSentry`, they're the ones calling it, without the parameterized route as a second parameter. We therefore need
+  // a different way to make it available, which we'll do by storing it on the handler or handler wrapper, depending on
+  // what we're given.
+  maybeWrappedHandler.__sentry_route__ = parameterizedRoute;
+
+  // In the simple case, where the handler has not yet been wrapped, `maybeWrappedHandler` will be passed to
+  // `withSentry` as `origHandler` and so the parameterized route will be easy to access.
+  if (!('__sentry_wrapped__' in maybeWrappedHandler)) {
+    return withSentry(maybeWrappedHandler);
+  }
+
+  // In the case where the exported handler is already of the form `withSentry(origHandler)`, `maybeWrappedHandler` is
+  // the wrapped handler returned by `withSentry`, which now has the parameterized route as a property on itself. By
+  // default in JS, functions have `global` as their `this` value, so to be able to get at the route value, we need to
+  // make sure it's called with itself as `this`. Since ultimately we need to give nextjs something *it* will call (and
+  // it won't set the `this` value we want when it does), this means one more wrapper.
+  const newWrapper: WrappedNextApiHandler = (req, res) => {
+    // Make `maybeyWrappedHandler` its own `this`
+    return maybeWrappedHandler.call(maybeWrappedHandler, req, res);
+  };
```

Comment:
I don't believe this is necessary since we also turned the function returned from `withSentry` from an arrow function into an actual `function`.

To confirm this, I changed this block to `return maybeWrappedHandler;` and quickly ran the integration tests you wrote, and they all seem to pass.

Am I missing something?

Reply:
Huh. My first step was going from arrow to regular function, for exactly this reason, and when I logged `this` inside of `withSentry`, it gave me `global`. It wasn't until I used `call` that logging `this` gave me the wrapped handler.

Lemme try it again to be sure.

Reply:
Ok got it. I am still confused as to how `this` works and I'll probably never get it. Always, when I think I have it, some thing comes up that completely nukes the concept I have of `this`.

Anyways, I somehow still don't like our current implementation for the following reasons:
- Messing with `this`/`call`/`bind` is always a bit funky.
- If the user wraps the API route with some other wrapper (e.g. `export dfault myWrapper(withSentry(handler))`) we double wrap the handler.

I then thought about doing the following which seems more or less bulletproof. We wrap the handler in any case with `withSentryAPI`, but we make `withSentry` idempotent by writing a flag to `req` when the request was handled by a `withSentry` wrapper, and bail in the beginning of `withSentry` when we see that `req` already has that flag. Wdyt?

Reply:
> Messing with this/call/bind is always a bit funky - also from a user perspective. Are we sure we're not stomping on user-provided this values?

With all due respect, I don't agree. Call, apply, and bind are perfectly respectable parts of the language and we use them as intended, all over the SDK. I will grant you that they're not used as commonly as some other language features, but that's exactly why I'm not especially worried about users having set their own `this` (generally, and especially in this particular situation, where there's nothing that lends itself to `this`-ing).

> If the user wraps the API route with some other wrapper (e.g. export dfault myWrapper(withSentry(handler))) we double wrap the handler.

This, however, is a good point. We (should, if we don't already) tell people to have us be on the outside, but I know as well as anyone from my time doing support that users don't always take our advice.

> I then thought about doing the following which seems more or less bulletproof: We wrap the handler in any case with withSentryAPI, but we make withSentry idempotent by writing a flag to req when the request was handled by a withSentry wrapper, and bail in the beginning of withSentry when we see that req already has that flag. Wdyt?

I think I wish we'd had this conversation before I spent hours in Higher Order Function Brain Melt City, but that's no one's fault but the time difference. I think it's a good solution to your second point, which I've now implemented.

Reply:
> Call, apply, and bind are perfectly respectable parts of the language and we use them as intended

Agree, but imo they are by far the most common source of confusion in our code base. So if we can, I would avoid them if possible. (I am aware that in many circumstances they are necessary - in these situations, they are fine ofc)